### DISCUSSION TITLE
Importing OPCUA.Di

### DISCUSSION THREAD

**Discussion by roin93** [FreeOpcUa/opcua-asyncio#498] [Q&A]
Hello,

I tried to import the OPCUA.Di Models and Nodes from [https://opcfoundation.org/UA/schemas/DI/1.2/](https://opcfoundation.org/UA/schemas/DI/1.2/)

```
server.import_xml("Opc.Ua.Di.NodeSet2.xml")
server.import_xml("OpcUaDiModel.xml")
```

Right at the beginning I get an Error that something is missing. Do I need to Import something else first?

> WARNING:asyncua.common.xmlparser:Parsing value of type 'QualifiedName' not implemented
> WARNING:asyncua.client.ua_client.UaClient:close_session was called but connection is closed
> WARNING:asyncua.client.ua_client.UaClient:close_secure_channel was called but connection is closed
> WARNING:asyncua.client.ua_client.UaClient:disconnect_socket was called but connection is closed
> Traceback (most recent call last):
>   File "main.py", line 272, in <module>
>     main()
>   File "main.py", line 210, in main
>     server.import_xml("OpcUaDiModel.xml")
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 94, in wrapper
>     result = self.tloop.post(aio_func(*args, **kwargs))
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 52, in post
>     return futur.result()
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 435, in result
>     return self.__get_result()
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
>     raise self._exception
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/server/server.py", line 562, in import_xml
>     return await importer.import_xml(path, xmlstring)
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlimporter.py", line 101, in import_xml
>     dnodes = self.parser.get_node_datas()
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlparser.py", line 166, in get_node_datas
>     node = self._parse_node(tag, child)
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlparser.py", line 181, in _parse_node
>     self._parse_attr(el, obj)
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlparser.py", line 224, in _parse_attr
>     self._parse_refs(el, obj)
>   File "/home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlparser.py", line 372, in _parse_refs
>     struct.reftype = ref.attrib["ReferenceType"]
> KeyError: 'ReferenceType'
> ERROR:asyncio:Task was destroyed but it is pending!
> task: <Task pending coro=<Client._renew_channel_loop() running at /home/msauser/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/client/client.py:398> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7feaa9943510>()]>>

I started building an opcua aggregator for servers generated by siemens plcs. Therefore I need to import plenty of DataTypes and such. 

Would be nice if someone knows that problem and can point me to the right files...

**Comment by swamper123** [comment]
The Opc.Ua.Di.NodeSet2.xml File from here:
https://github.com/OPCFoundation/UA-Nodeset/tree/v1.04/DI

This imports the DI spec. 

The second file you import (seems to be a self made one from you?), seems to miss a ReferenceType for a Reference in the xml-File. This blows things up.

**Reply by roin93** [reply]
The second file is from the Webpage I linked.

**Reply by swamper123** [reply]
Afaik, you should only need the first one.

**Reply by roin93** [reply]
Good information that I only need the first one.

Nevertheless I replaced the file I had before by the linked file from you:

> Importing opcua.di
> WARNING:asyncua.client.ua_client.UaClient:close_session was called but connection is closed
> WARNING:asyncua.client.ua_client.UaClient:close_secure_channel was called but connection is closed
> WARNING:asyncua.client.ua_client.UaClient:disconnect_socket was called but connection is closed
> Traceback (most recent call last):
>   File "main_async_sync.py", line 276, in <module>
>     main()
>   File "main_async_sync.py", line 211, in main
>     server.import_xml("Opc.Ua.Di.NodeSet2.xml")
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 94, in wrapper
>     result = self.tloop.post(aio_func(*args, **kwargs))
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 52, in post
>     return futur.result()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 435, in result
>     return self.__get_result()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
>     raise self._exception
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/server/server.py", line 562, in import_xml
>     return await importer.import_xml(path, xmlstring)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlimporter.py", line 95, in import_xml
>     await self._check_required_models(xmlpath, xmlstring)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlimporter.py", line 64, in _check_required_models
>     req_models = self.parser.list_required_models(xmlpath, xmlstring)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/xmlparser.py", line 396, in list_required_models
>     tree = ET.parse(xmlpath)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/xml/etree/ElementTree.py", line 1197, in parse
>     tree.parse(source, parser)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/xml/etree/ElementTree.py", line 598, in parse
>     self._root = parser._parse_whole(source)
> xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 71, column 88
> ERROR:asyncio:Task was destroyed but it is pending!
> task: <Task pending coro=<Client._renew_channel_loop() running at /home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/client/client.py:398> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f27339f25d0>()]>>

**Reply by swamper123** [reply]
DI should be importable as is...

Can you provide more information?
- Python (3.7 as I see)
- What is your code looking?
- Where is your file located in your project hierarchy? (is it in the same folder as your python code or somewhere else?)

**Reply by roin93** [reply]
As you saw, I am using python 3.7.

The relevant part of the code is lookig as following:
```
import logging
import asyncio
from asyncua.sync import Client
from asyncua.sync import Server as opcuaServer
from asyncua import ua
import socket
import threading
import time

from asyncua.ua import NodeClass, LocalizedText, QualifiedName
from asyncua.sync import ThreadLoop

def main():
    logging.basicConfig(level=logging.WARN)
    # logger = logging.getLogger("KeepAlive")
    # logger.setLevel(logging.DEBUG)
    address = "opc.tcp://192.168.178.2:4840"
    # client = Client("opc.tcp://localhost:4840/freeopcua/server/")
    with ThreadLoop() as tloop:
        with Client(address, tloop=tloop) as client:
            # client = Client("opc.tcp://admin@localhost:4840/freeopcua/server/") #connect using a user

            # eigenen server starten
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(("8.8.8.8", 80))
            ip_addr = s.getsockname()[0]

            server = opcuaServer(tloop=tloop)
            server.set_endpoint("opc.tcp://%s:4840/opcua/" % ip_addr)
            server.set_server_name("OwnServer")
            print("server online")

            try:
                client.connect()

                # Client has a few methods to get proxy to UA nodes that should always be in address space such as Root or Objects
                # root = client.get_root_node()
                root = client.nodes.root
                print("Root node is: ", root)
                # objects = client.get_objects_node()
                objects = client.nodes.objects
                print("Objects node is: ", objects)

                # Node objects have methods to read and write node attributes as well as browse or populate address space
                children = root.get_children()
                print("Children of root are: ", children)

                # server_root = server.get_root_node()
                server_root = server.nodes.root
                # server_objects = server.get_objects_node()
                server_objects = server.nodes.objects

                # Namespaces
                # 0: Default opcua namespace http://opcfoundation.org/UA/
                # 1: freeopcua / urn:freeopcua:python:server
                # 2: opcua.di http://opcfoundation.org/UA/DI/
                # 3: simatic http://www.siemens.com/simatic-s7-opcua

                # Add Datatype of the opcua.di specs

                if True: # Im just testing out different parts of the code and use this kind for disabling blocks. If I comment it out, I will remove it...
                    print("Importing opcua.di")
                    server.import_xml("Opc.Ua.Di.NodeSet2.xml")
                    # server.import_xml("OpcUaDiModel.xml")

                    print("Imported opcua.di")
                else:
                    server.register_namespace("http://opcfoundation.org/UA/DI/")

                # Some other stuff...
if __name__ == "__main__":
    main()
```

The file is located in the same folder as my main.py

**Reply by swamper123** [reply]
I am not very familiar with Threadloop (and the whole sync stuff, I just know that it exist).
I think you are missing server.init() (after server = opcuaServer(tloop=tloop)) as well?

Still wondering why the parser has a problem to parse this xml file ....

**Reply by roin93** [reply]
The init function for the (sync) Server class does not exist. Therefore I guess it is not needed, besides the fact I never used it.

I currently parsing my old xml file, just getting the Warning, that parsing QualifiedName is not implemented.

**Reply by roin93** [reply]
The mentioned warning is `WARNING:asyncua.common.xmlparser:Parsing value of type 'QualifiedName' not implemented` when I try to use `server.import_xml(my_linked_xml)`.
When I try to load the file you mentioned I get the other errors I cannot work around at the moment.

**Reply by swamper123** [reply]
> 
> 
> The init function for the (sync) Server class does not exist. Therefore I guess it is not needed, besides the fact I never used it.
> 

You are right. It is already executed in the __init__:
https://github.com/FreeOpcUa/opcua-asyncio/blob/0a1cce0783d93493515ff92e55952e63909fc834/asyncua/sync.py#L248

> The mentioned warning is WARNING:asyncua.common.xmlparser:Parsing value of type 'QualifiedName' not implemented

Yes, that is correct, we have some missing values:
https://github.com/FreeOpcUa/opcua-asyncio/blob/0a1cce0783d93493515ff92e55952e63909fc834/asyncua/common/xmlparser.py#L304

```python
            # Missing according to string_to_val: XmlElement, ExpandedNodeId,
            # QualifiedName, StatusCode.
            # Missing according to ua.VariantType (also missing in string_to_val):
            # DataValue, Variant, DiagnosticInfo.
```
@oroulet Do you know, if we support some of these mentioned values now? Git Blame says, it's a while ago since somebody touched these lines. If yes, we may provide an update on that.
